Dev

Bivis63 opened the pull request.

Bivis63 committed:
дз спринт 23 (реализовал экран плейлиста с отображением списка треков и возможностью удалять треки)

Bivis63 committed:
дз спринт 23 (установил новую иконку , реализовал кнопки share и menu, реализовал удаление  и редактирование плейлиста)

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/data/impl/playlistsimpl/PlayListsRepositoryImpl.kt:
```diff
@@ -59,11 +58,71 @@ class PlayListsRepositoryImpl(
+        )
+    }
+
+    override suspend fun getAllPlaylistTracks(playlistId: List<Long>): List<Track> {
```
Привет! В этом аргументе хранится список `id`, поэтому лучше назвать её например `playListIds` для лучшей читаемости кода.

Bivis63 replied:
Привет , незнаю куда написать ,скину ещё менторам , проверил склонение треков у меня и на эмуляторе и на телефоне количество треков склоняеться как нужно ,видео пришлю в пачку :)

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/data/impl/playlistsimpl/PlayListsRepositoryImpl.kt:
```diff
@@ -59,11 +58,71 @@ class PlayListsRepositoryImpl(
+
+    override suspend fun getAllPlaylistTracks(playlistId: List<Long>): List<Track> {
+        val playlist = appDatabasePlayLists.getPlayListsDao().getAllPlaylistTracks()
+        val trackList = arrayListOf<Track>()
```
чтобы не создавать лишних списков руками и сделать код короче рекомендую ознакомиться с функциями-расширениями `map` и `filter`.
Смотри как коротко можно переписать этот же код

```
        val playlist = appDatabasePlayLists.getPlayListsDao()
            .getAllPlaylistTracks()
        return playlist
            .filter { it.trackId.toLong() in playlistId }
            .map { convertFromTrackEntity(it) } 
```

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/domain/db/playlists/PlayListsInteractor.kt:
```diff
@@ -18,5 +19,14 @@ interface PlayListsInteractor {
 
     fun saveImageToPrivateStorage(uri: Uri, context: Context):Uri?
```
сохранение картинки это операция ввода вывода, которая может занимать много времени. Настоятельно рекомендую сделать эту функцию suspend-функцией как и остальные функции в этом интеракторе. В репозитории тоже следует поправить.

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/editPlaylist/EditPlaylistFragment.kt:
```diff
@@ -0,0 +1,143 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+
+        playList = arguments?.getSerializable("playlistModel") as? PlayListsModels
```
рекомендую ключи всех аргументов выносить в companion object фрагмента. Так будет проще ориентироваться в коде и ты не допустишь ошибок, если потребуется изменять имена аргументов

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/editPlaylist/EditPlaylistFragment.kt:
```diff
@@ -0,0 +1,143 @@
+
+class EditPlaylistFragment : NewPlayListFragment() {
+
+    private var binding: FragmentEditPlaylistBinding? = null
```
удобнее сделать shadow поле 
`private var_binding: FragmentEditPlaylistBinding? = null`
и поле для доступа к биндингу
`private val binding: FragmentEditPlaylistBinding get() = _binding!!`
так будет удобнее и не придется писать `!!` при каждом обращении к биндингу.
Так же необходимо обнулять binding = null в методе `onDestroyView()`, потому что андрид может пересоздать фрагмент, и при это view фрагмента пересоздастся. Поэтому нельзя хранить ссылку на старый `viewbinding`

vlad-zotov replied:
я посмотрел по проекту, ты везде используешь `lateinit var` для работы с `viewbinding`. Это неправильно. Необходимо использовать нуллабельнный тип `private var binding: SomeBinding?` чтобы была возможность обнулять `viewbinding` в `onDestroyView()`. Иначе ты получишь утечку памяти в случае пересоздания фрагмента.

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/playlistscreen/PlaylistScreenFragment.kt:
```diff
@@ -0,0 +1,323 @@
+
+        viewModel.observeDurationState().observe(viewLifecycleOwner) {
+            if (it != null) {
+                val duration = SimpleDateFormat("mm", Locale.getDefault()).format(it).toInt()
```
Рекомендую весь код, связанный с форматированием времени и дат вынести в отдельный класс-утилиту `object DateTimeUtil`. Это позволяет избавиться от дублирования кода, так же удобно следить за всеми форматами дат и времени в проекте. Ещё этот класс будет легко перенести в другой проект.

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/playlistscreen/PlaylistScreenFragment.kt:
```diff
@@ -0,0 +1,323 @@
+            if (it != null) {
+                val duration = SimpleDateFormat("mm", Locale.getDefault()).format(it).toInt()
+                val formattedDuration =
+                    resources.getQuantityString(R.plurals.minutes, duration, duration)
```
очень хорошо, что используешь `plurals`!

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/playlistscreen/PlaylistScreenFragment.kt:
```diff
@@ -0,0 +1,323 @@
+            openDialogDeletePlaylist()
+        }
+        val bundle = Bundle().apply {
+            putSerializable("playlistModel", playlistModel)
```
`"playlistModel"` стоит вынести в константы и переиспользовать

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/media/ui/playlistscreen/PlaylistScreenFragment.kt:
```diff
@@ -0,0 +1,323 @@
+            .setPositiveButton(resources.getString(R.string.delete)) { dialog, which ->
+                viewModel.deleteTrackFromPlaylist(trackId)
+                val playListsFragment =
+                    parentFragmentManager.findFragmentByTag("PlayListsFragment") as? PlayListsFragment
```
так лучше не делать, потому что ты не можешь гарантировать в каком состоянии у тебя `PlayListsFragment`, который в текущий момент не отображен юзеру. Могу предложить здесь два способа обновлять фрагмент с плейлистами. Первый - обновлять `PlayListsFragment` в методе `onResume()`, который вызывается перед тем как отобразить фрагмент юзеру. Второй - общая `ViewModel` находящаяся в actiivty. https://developer.android.com/guide/fragments/communicate#viewmodel здесь подробнее

vlad-zotov commented on app/src/main/res/layout/fragment_edit_playlist.xml:
```diff
@@ -0,0 +1,104 @@
+        android:id="@+id/backToMedia"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:title="Редактировать" />
```
строки необходимо вынести в ресурсы

vlad-zotov commented on app/src/main/res/layout/fragment_edit_playlist.xml:
```diff
@@ -0,0 +1,104 @@
+        android:layout_width="0dp"
+        android:layout_height="0dp"
+        app:layout_constraintDimensionRatio="1:1"
+        android:layout_marginStart="24dp"
```
размеры и отступы необходимо вынести в ресурсы и переиспользовать

vlad-zotov commented on app/src/main/res/layout/fragment_edit_playlist.xml:
```diff
@@ -0,0 +1,104 @@
+        android:backgroundTint="@color/grey"
+        android:enabled="false"
+        android:padding="@dimen/padding_10dp"
+        android:text="Сохранить"
```
строки необходимо вынести в ресурсы

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="#E6E8EB"
```
цвет необходимо вынести в ресурсы

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/margin_start_16dp"
+            android:layout_marginTop="@dimen/margin_top_24dp"
+            android:text="Best Song 2021"
```
используй здесь префикс `tools`, тогда этот текст будет видно только на превью и он не попадет в реальный проект
`tools:text="Best Song 2021"`
Избегай использования такого хардкода как сейчас в проектах. Рано или поздно это протекает в релиз и это вызывает негативный отклик!

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/margin_start_16dp"
+            android:layout_marginTop="@dimen/margin_top_8dp"
+            android:text="2022"
```
аналогично предыдущему комментарию

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/margin_start_16dp"
+            android:layout_marginTop="@dimen/margin_top_8dp"
+            android:text="300 минут"
```
аналогично предыдущему комментарию

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+            android:id="@+id/trackCount"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="98 треков"
```
аналогично предыдущему комментарию

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+    <LinearLayout
+        android:id="@+id/menu_bottom_sheet"
+        style="@style/bottomSheetStyle"
+        app:behavior_peekHeight="383dp"
```
следует вынести в ресурсы

vlad-zotov commented on app/src/main/res/layout/fragment_playlist_screen.xml:
```diff
@@ -0,0 +1,236 @@
+                    android:layout_marginStart="@dimen/margin_start_8dp"
+                    android:layout_marginTop="@dimen/margin_top_6dp"
+                    android:fontFamily="@font/ys_display_regular"
+                    android:text="Best Song 2021"
```
весь хардкод для верстки пометь префиксом `tools` Это очень удобно! Бережет время и нервы :)

Bivis63 committed:
дз спринт 23
1.Убрал баг в котором при нажатии на кнопку Cancel в диалоге удаленичя трека открывается меню плейлиста.
2.Треки склоняються как нужно видео отправлю в пачку
3.Попровил обложку ,placeHolder и сделал привязку bottomSheet в процентах
4.В диалоговом окне покрасил кнопки и фон
5.использовал в качестве обложки artworkUrl60
6.Вынес в ресурсы всё что было хардкодом и для отображения сделал tools
7.Функцию сохранения картинки сделал suspend
